internal/gengapic/genrest_test.go

```diff
@@ -0,0 +1,231 @@
+// Copyright (C) 2021  Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gengapic
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/protoc-gen-go/descriptor"
+	plugin "github.com/golang/protobuf/protoc-gen-go/plugin"
+	"github.com/googleapis/gapic-generator-go/internal/pbinfo"
+	"google.golang.org/genproto/googleapis/api/annotations"
+)
+
+// This is a hack to provide an lvalue for a constant.
+func typep(typ descriptor.FieldDescriptorProto_Type) *descriptor.FieldDescriptorProto_Type {
+	return &typ
+}
+
+// Another hack, this time for field numbers.
+func idxAddr(i int) *int32 {
+	j := int32(i)
+	return &j
+}
+
+func setupMethod(t *testing.T, g *generator, url string, allFieldNames []string) *descriptor.MethodDescriptorProto {
```

Comment:
My rationale for using `allFieldNames` instead of `fields` is that I wanted to make it clear that this slice contained _all_ the field names for the message, not just those that were path or query params. The same info can be transmitted via comments, so I'll try renaming and commenting.